refactor: Add consteval uint256 constructor

hodlinator opened the pull request:
Motivation:
* Validates and converts the hex string at compile time instead of at runtime into the resulting bytes.
* Makes it possible to derive other compile time constants from `uint256`.
* Potentially eliminates runtime dependencies (`SetHexDeprecated()` is called in less places).
* Has stricter requirements than the deprecated `uint256S()` (requiring 64 chars exactly, disallows garbage at the end) and replaces it in a bunch of places.
* Makes the binary smaller (tested Guix-built x86_64-linux-gnu bitcoind binary).
* Minor: should shave off a few cycles of start-up time.

Extracted from #30377 which diverged into exploring `consteval` `ParseHex()` solutions.

hodlinator commented:
> > Eliminates runtime dependencies.
> 
> Is this true? IIRC it was true for `ParseHex`, because that returns a `std::vector<std::byte>` (and the module that was used to create the bytes could be dropped before linking. However, the `uint256` type lives in the same module as the parsing code, so I don't think this is true?

Your memory serves you right: https://github.com/bitcoin/bitcoin/pull/30377#issuecomment-2214432177
However, this change eliminates some calls to the .cpp `SetHexDeprecated()` which has potential to decrease runtime dependencies in the future, even if it were not to happen right now. Adjusted summary.

> 
> > Has stricter requirements than the deprecated uint256S() and replaces it in a bunch of places.
> 
> It would be good to mention them: "Requires exactly 64 hex characters" (or so). Also, it could recap the risks of uint256S: "Previously, any invalid string was accepted silently."?

Cheers, adjusted summary somewhat.

maflcko reviewed (approved):
lgtm. Left some nits, feel free to ignore them.

Mazzika1 reviewed (commented):
Mazzika

Mazzika1 reviewed (changes requested):
Mazzika

hodlinator commented on src/test/arith_uint256_tests.cpp:
```diff
@@ -22,7 +22,7 @@ static inline arith_uint256 arith_uint256V(const std::vector<unsigned char>& vch
     return UintToArith256(uint256(vch));
 }
-static inline arith_uint256 arith_uint256S(const std::string& str) { return UintToArith256(uint256S(str)); }
+static inline arith_uint256 arith_uint256S(std::string_view str) { return UintToArith256(uint256S(str)); }
```
Keeping as is unless anyone feels stronger about this. Not big enough to separate out to own PR IMO.

hodlinator replied:
Dropped in latest force-push. #30526 is the true follow-up change for #30436 and already contains a better version of it.

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -307,7 +307,7 @@ BOOST_AUTO_TEST_CASE(parse)
         std::string s_12{"0000000000000000000000000000000000000000000000000000000000000012"};
         BOOST_CHECK_EQUAL(uint256S("12\0").GetHex(), s_12);
-        BOOST_CHECK_EQUAL(uint256S(std::string{"12\0", 3}).GetHex(), s_12);
+        BOOST_CHECK_EQUAL(uint256S(std::string_view{"12\0", 3}).GetHex(), s_12);
```
Same https://github.com/bitcoin/bitcoin/pull/30560#discussion_r1699068420

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -336,4 +336,13 @@ BOOST_AUTO_TEST_CASE( check_ONE )
+{
+    constexpr uint256 consteval_uint{
+           "4a5e1e4baab89f3a32518a88c31bc87f618F76673E2CC77AB2127B7AFDEDA33B"};
+    const uint256 runtime_uint = uint256S(std::string_view{
```
Thanks, shows that this change predates `FromHex`. :) Fixed!

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+template<unsigned int BITS>
+consteval base_blob<BITS>::base_blob(std::string_view hex_str)
+{
+    // Non-lookup table version of HexDigit().
```
I'd rather hold off on breaking it out until this one is merged, not sure the `ParseHexConsteval` header will be including this one.

hodlinator commented on src/uint256.h:
```diff
@@ -127,6 +150,7 @@ class uint160 : public base_blob<160> {
 class uint256 : public base_blob<256> {
 public:
     static std::optional<uint256> FromHex(std::string_view str) { return detail::FromHex<uint256>(str); }
+    consteval explicit uint256(std::string_view hex_str) : base_blob<256>(hex_str) {}
```
Attempted doing:
```C++
using base_blob<256>::base_blob;
```
and same for `uint160`.

But ran into GCC issues with inherited consteval constructors (didn't try other compilers): https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106084

Tried adding
```C++
consteval explicit uint160(std::string_view hex_str)
```
And replacing
```C++
    BOOST_CHECK_EQUAL(HexStr(cc1.out.scriptPubKey), HexStr(GetScriptForDestination(PKHash(uint160(ParseHex("816115944e077fe7c803cfa57f29b36bf87c1d35"))))));
```
with
```C++
    BOOST_CHECK_EQUAL(HexStr(cc1.out.scriptPubKey), HexStr(GetScriptForDestination(PKHash(uint160{"816115944e077fe7c803cfa57f29b36bf87c1d35"})))));
```
But test failed because the bytes are reversed. So holding off on touching `uint160`.

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -153,8 +153,8 @@ BOOST_AUTO_TEST_CASE( comparison ) // <= >= < >
-    BOOST_CHECK_LT(uint256S("2000000000000000000000000000000000000000000000000000000000000001"),
-                   uint256S("1000000000000000000000000000000000000000000000000000000000000002"));
+    BOOST_CHECK_LT(uint256{"2000000000000000000000000000000000000000000000000000000000000001"},
+                   uint256{"1000000000000000000000000000000000000000000000000000000000000002"});
```
Good point, fixed by changing regexp:

`sed -i --regexp-extended -e 's/uint256S\(\"(0x)?([^\"]*)\"\)/uint256{\"\2\"}/g' $(git grep -l 'uint256S' -- ':src' ':(exclude)src/test/uint256_tests.cpp' ':(exclude)src/test/arith_uint256_tests.cpp')`
->
`sed -i --regexp-extended -e 's/uint256S\(\"(0x)?([^\"]{64})\"\)/uint256{\"\2\"}/g' $(git grep -l 'uint256S' -- ':src')`

hodlinator replied:
Thanks, that helped narrow it down a bit:
``sed -i --regexp-extended -e 's/uint256S\(\"(0x)?([^\"]{64})\"\)/uint256{\"\2\"}/g' $(git grep -l 'uint256S' -- ':src')``
->
``sed -i --regexp-extended -e 's/uint256S\("(0x)?([^"]{64})"\)/uint256{"\2"}/g' $(git grep -l uint256S)``

maflcko reviewed (approved):
Left some more nits.

ACK 9315c1f34abaa4cd78185818ed8d2df5f35eb3ce 🕵

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 9315c1f34abaa4cd78185818ed8d2df5f35eb3ce 🕵
GZQfOWL8G0spNKvjqmkS65RYUYIQBpZjUi0j+fcqLs9A5d+ffS+M/zKWMvJBpjmBad5mU4mrr/6AY/sNvV6DBg==
```

</details>

hodlinator commented on src/test/uint256_tests.cpp:
```diff
@@ -336,4 +336,11 @@ BOOST_AUTO_TEST_CASE( check_ONE )
+BOOST_AUTO_TEST_CASE(FromHex_vs_uint256)
+{
+    const uint256 runtime_uint = uint256::FromHex("4A5E1E4BAAB89F3A32518A88C31BC87F618f76673e2cc77ab2127b7afdeda33b").value();
+    constexpr uint256 consteval_uint{             "4a5e1e4baab89f3a32518a88c31bc87f618F76673E2CC77AB2127B7AFDEDA33B"};
```
Yeah, should probably try to incorporate `clang-format` more into my changes.

Made the indentation in `FromHex_vs_uint256` a bit less insane by hand without introducing new variables.

Fixed missing space after `template`.

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+    auto from_hex = [](const char c) -> int8_t {
+        if (c >= '0' && c <= '9') return c - '0';
+        if (c >= 'a' && c <= 'f') return c - ('a' - 0xA);
+        if (c >= 'A' && c <= 'F') return c - ('A' - 0xA);
```
Fixed.

sedited reviewed (approved):
ACK ed9823f7175a451a0170490e4eb96ceb2632b7a0

sedited reviewed (approved):
Re-ACK c0b259b62a16e5aa657b817d99167e4295342ca5

Only changes are dropping the commit changing some strings to string_view.

hodlinator commented:
@TheCharlatan thanks for the speedy re-ACK! My latest force-push actually dropped a tangential commit: https://github.com/bitcoin/bitcoin/pull/30560#discussion_r1700106826

stickies-v reviewed (approved):
ACK c0b259b62a16e5aa657b817d99167e4295342ca5

Very nice PR, more compile-time checks and easy to review. Makes my branch removing `uint256S` a bunch more compact too. Non-blocking nits.

paplorinc commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+        assert(false);
+    };
+
+    // 2 chars per byte.
```
The purpose of code should be to express context which we couldn't convey using code, right?
Here the code already states this clearly, the comment just seems like a distraction.

hodlinator replied:
Slightly more okay if moved to the same line? https://github.com/bitcoin/bitcoin/pull/30560#discussion_r1700885770

paplorinc replied:
it's just a nit from my part, but I just see noise here, I'm not a big fan of dead comments, especially when they repeat the code

hodlinator replied:
Agree it's better to make the code self-documenting in general. My threshold for what is a useless comment seems slightly different than yours in this case.

l0rinc reviewed (commented):
I've reviewed a part of it, left some questions.
I will have to measure the before/after compilation speeds and the before/after artefact size to be able to approve - will do that in the following days.

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+    };
+
+    // 2 chars per byte.
+    assert(hex_str.length() == m_data.size() * 2);
```
Seeing as `m_data` is used two lines below, I prefer mentioning it here.

I will however move the comment down to this line like you suggested for the other `assert`.

hodlinator committed:
refactor: Add consteval uint256(hex_str)

Complements uint256::FromHex() nicely in that it naturally does all error checking at compile time and so doesn't need to return an std::optional.

Will be used in the following 2 commits to replace many calls to uint256S(). uint256S() calls taking C-string literals are littered throughout the codebase and executed at runtime to perform parsing unless a given optimizer was surprisingly efficient. While this may not be a hot spot, it's better hygiene in C++20 to store the parsed data blob directly in the binary, without any parsing at runtime.

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+template <unsigned int BITS>
+consteval base_blob<BITS>::base_blob(std::string_view hex_str)
+{
+    // Non-lookup table version of HexDigit().
```
Was having issues with inefficient MSVC code-generation and also for a time was trying to make it `constexpr` instead of `consteval` and missing `static constexpr` from C++23. It's doable under `consteval` but the diff ain't too pretty:
```diff
diff --git a/src/uint256.h b/src/uint256.h
index a52b0a1f71..be27be93de 100644
--- a/src/uint256.h
+++ b/src/uint256.h
@@ -98,18 +98,35 @@ consteval base_blob<BITS>::base_blob(std::string_view hex_str)
 {
     // Non-lookup table version of HexDigit().
     auto from_hex = [](const char c) -> int8_t {
-        if (c >= '0' && c <= '9') return c - '0';
-        if (c >= 'a' && c <= 'f') return c - 'a' + 0xA;
-        if (c >= 'A' && c <= 'F') return c - 'A' + 0xA;
-
-        assert(false); // Reached if ctor is called with an invalid hex digit.
+        constexpr signed char hexdigit[256] =
+        { -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          0,1,2,3,4,5,6,7,8,9,-1,-1,-1,-1,-1,-1,
+          -1,0xa,0xb,0xc,0xd,0xe,0xf,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,0xa,0xb,0xc,0xd,0xe,0xf,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
+          -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1, };
+
+        return hexdigit[(uint8_t)c];
     };
 
     assert(hex_str.length() == m_data.size() * 2); // 2 hex digits per byte.
     auto str_it = hex_str.rbegin();
     for (auto& elem : m_data) {
         auto lo = from_hex(*(str_it++));
-        elem = (from_hex(*(str_it++)) << 4) | lo;
+        assert(lo != -1);
+        auto hi = from_hex(*(str_it++));
+        assert(hi != -1);
+        elem = (hi << 4) | lo;
     }
 }
```
I'll see if I can get some measurements on compile times, if it becomes a factor, I'll reconsider this.

paplorinc replied:
useful context, thanks!

hodlinator commented on src/kernel/chainparams.cpp:
```diff
@@ -76,17 +86,17 @@ class CMainParams : public CChainParams {
         consensus.nSubsidyHalvingInterval = 210000;
         consensus.script_flag_exceptions.emplace( // BIP16 exception
-            uint256S("0x00000000000002dc756eebf4f49723ed8d30cc28a5f108eb94b1ba88ac4f9c22"), SCRIPT_VERIFY_NONE);
+            uint256{"00000000000002dc756eebf4f49723ed8d30cc28a5f108eb94b1ba88ac4f9c22"}, SCRIPT_VERIFY_NONE);
```
TODO: Measure.

paplorinc replied:
I also don't think this will be significant, will rent a cheap hetzner server and run something like:
```
hyperfine \
  --export-markdown make_bench.md \
  --warmup 1 --runs 10 \
  --parameter-list COMMIT 30cef5,be750d \
  --prepare 'git checkout {COMMIT} && git clean -fxd && git reset --hard && ./autogen.sh && ./configure --disable-ccache' \
  'echo {COMMIT} && make -j1'
  ```
  
i.e. 10 single-threaded, uncached compilations for 30cef5 vs be750d - would that be representative?

fanquake replied:
> uncached compilations for 30cef5 vs be750d - would that be representative?

Maybe, but I don't think this is that useful. Anyone doing development, should be using `ccache`, only compiling the code they need, etc, and in that case, I don't think compilation is "way too slow". Anyone who isn't doing development, is likely only compiling once, where +/- N seconds/minutes in compilation time, doesn't really matter.

paplorinc replied:
```
Summary
  'echo 30cef5 && make -j1 && ./src/bitcoind --version | head -n 1' ran
    1.00 ± 0.00 times faster than 'echo be750d && make -j1 && ./src/bitcoind --version | head -n 1'
```
both of them take 45 minutes on a i7-7700, without cache.

hodlinator replied:
Thanks @paplorinc for helping put that worry to rest!

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+    // 2 chars per byte.
+    assert(hex_str.length() == m_data.size() * 2);
+    auto str_it = hex_str.rbegin();
+    for (auto& elem : m_data) {
+        auto lo = from_hex(*(str_it++));
+        elem = (from_hex(*(str_it++)) << 4) | lo;
+    }
```
Yes, that is why I inlined `lo`. I don't see the danger now that `str_it` is only used once per statement. You used `--it` in your suggestion where you only broke out one of the parts: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686328446

paplorinc replied:
It's a nit, I'm just providing an alternative, will of course let you decide

hodlinator commented on src/uint256.h:
```diff
@@ -91,6 +93,27 @@ class base_blob
+    // Non-lookup table version of HexDigit().
+    auto from_hex = [](const char c) -> int8_t {
+        if (c >= '0' && c <= '9') return c - '0';
+        if (c >= 'a' && c <= 'f') return c - 'a' + 0xA;
```
Feel this has already been bikeshedded enough today. https://github.com/bitcoin/bitcoin/pull/30560#discussion_r1699604279

paplorinc replied:
Maybe, but was my first bikeshedding attempt at it :)

hodlinator commented on src/kernel/chainparams.cpp:
```diff
@@ -197,15 +207,15 @@ class CTestNetParams : public CChainParams {
         consensus.nSubsidyHalvingInterval = 210000;
         consensus.script_flag_exceptions.emplace( // BIP16 exception
-            uint256S("0x00000000dd30457c001f4095d208cc1296b0eed002427aa599874af7a432b105"), SCRIPT_VERIFY_NONE);
+            uint256{"00000000dd30457c001f4095d208cc1296b0eed002427aa599874af7a432b105"}, SCRIPT_VERIFY_NONE);
```
https://github.com/bitcoin/bitcoin/pull/30560#discussion_r1700907290

paplorinc commented on src/test/uint256_tests.cpp:
```diff
@@ -332,8 +332,15 @@ BOOST_AUTO_TEST_CASE(parse)
 
+BOOST_AUTO_TEST_CASE(FromHex_vs_uint256)
+{
+    auto runtime_uint{uint256::FromHex("4A5E1E4BAAB89F3A32518A88C31BC87F618f76673e2cc77ab2127b7afdeda33b").value()};
```
are we sure we want to mandate that mixed-case hex strings can be parsed?

paplorinc replied:
Thanks for considering, I'm personally fine with doing it in a different PR as well

stickies-v replied:
I don't have a strong view on this but I would prefer getting this PR over the line as is.

hodlinator replied:
Trying to not spawn yet another PR from this work but will post that as a separate PR then if this one is merged in the current state.

TheCharlatan commented on src/uint256.h:
```diff
@@ -91,6 +93,26 @@ class base_blob
+        if (c >= 'a' && c <= 'f') return c - 'a' + 0xA;
+        if (c >= 'A' && c <= 'F') return c - 'A' + 0xA;
+
+        assert(false); // Reached if ctor is called with an invalid hex digit.
```
Might be a good idea to add a small suggestion in the developer-notes for using `throw` for these checks, since we already do so in https://github.com/bitcoin/bitcoin/blob/master/src/script/miniscript.h#L177 . I initially assumed this was not supported, but seems like that was fixed in recent compiler versions.

hodlinator replied:
Iff I retouch, I might experiment with including the offending digit in the `throw`.

hodlinator replied:
Yeah, I experimented with using `std::format` but at least GCC would not print the actual thrown value anyway.

stickies-v commented:
re-ACK be750dccb08050bcca48c4236ca6b902dd911c7b - doc-only changes to address review nits

sedited reviewed (approved):
Re-ACK be750dccb08050bcca48c4236ca6b902dd911c7b

paplorinc commented:
ACK be750dccb08050bcca48c4236ca6b902dd911c7b

hodlinator committed:
scripted-diff: Replace uint256S("str") -> uint256{"str"}

-BEGIN VERIFY SCRIPT-
sed -i --regexp-extended -e 's/\buint256S\("(0x)?([^"]{64})"\)/uint256{"\2"}/g' $(git grep -l uint256S)
-END VERIFY SCRIPT-

hodlinator commented:
Rebased with minimal changes to fix merge-issues.

``sed -i --regexp-extended -e 's/uint256S\("(0x)?([^"]{64})"\)/uint256{"\2"}/g' $(git grep -l uint256S)``
->
``sed -i --regexp-extended -e 's/\buint256S\("(0x)?([^"]{64})"\)/uint256{"\2"}/g' $(git grep -l uint256S)``
to make it avoid matching new instances of `arith_uint256S` being passed 64 chars for the first time in **arith_uint256_tests.cpp**.

paplorinc commented:
ACK a34ba0eed401f2b48e168b857d05eab065d9ab42

hodlinator committed:
refactor: Hand-replace some uint256S -> uint256

chainparams.cpp - workaround for MSVC bug triggering C7595 - Calling consteval constructors in initializer lists fails, but works on GCC (13.2.0) & Clang (17.0.6).

l0rinc reviewed (approved):
ACK 2d9d752e4fc30aabf2ddd36ca7a63432d26d4fea

stickies-v commented:
re-ACK 2d9d752e4fc30aabf2ddd36ca7a63432d26d4fea
